crates/winch/src/codegen.rs

```diff
@@ -0,0 +1,221 @@
+use crate::{
+    abi::{align_to, local::LocalSlot, ABISig},
+    frame::Frame,
+    masm::{MacroAssembler, OperandSize, RegImm},
+    regalloc::RegAlloc,
+    stack::Stack,
+};
+use anyhow::Result;
+use wasmparser::{FuncValidator, FunctionBody, ValidatorResources};
+use wasmtime_environ::WasmType;
+
+/// The code generation context
+pub(crate) struct CodeGenContext<'a, M>
+where
+    M: MacroAssembler,
+{
+    pub masm: M,
+    pub stack: Stack,
+    pub frame: &'a Frame,
+}
+
+impl<'a, M> CodeGenContext<'a, M>
+where
+    M: MacroAssembler,
+{
+    pub fn new(masm: M, stack: Stack, frame: &'a Frame) -> Self {
+        Self { masm, stack, frame }
+    }
+}
+
+/// The code generation abstraction
+pub(crate) struct CodeGen<'c, 'a: 'c, M>
+where
+    M: MacroAssembler,
+{
+    /// A reference to the function body
+    function: &'c mut FunctionBody<'a>,
+
+    /// The word size information, extracted from the current ABI
+    word_size: u32,
+
+    /// The ABI-specific representation of the function signature, excluding results
+    sig: ABISig,
+
+    /// The code generation context
+    pub context: CodeGenContext<'c, M>,
+
+    /// The register allocator
+    pub regalloc: RegAlloc,
+
+    /// Function body validator
+    pub validator: &'a mut FuncValidator<ValidatorResources>,
+}
+
+impl<'c, 'a: 'c, M> CodeGen<'a, 'c, M>
+where
+    M: MacroAssembler,
+{
+    pub fn new(
+        context: CodeGenContext<'c, M>,
+        word_size: u32,
```

Comment:
`word_size` should probably be obtained from the `MacroAssembler` or other platform-specific trait?